### PR TITLE
boards: initial import of Adafruit Feather nRF52840 Express

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
This provides support initial support for the [Adafruit Feather nRF52840 Express][feather-nrf52840]. The current feature set includes:

- Default flashing via Segger J-Link
- UART, SPI, and I2C exposed in accordance to the Feather [pinout]
- SAUL support for on-board LEDs and button
- `stdio_cdc_acm` using the [RIOT standard peripheral PID][pid.codes]

[feather-nrf52840]: https://learn.adafruit.com/introducing-the-adafruit-nrf52840-feather/
[pinout]: https://cdn-learn.adafruit.com/assets/assets/000/046/243/original/adafruit_products_Feather_M0_Adalogger_v2.2-1.png
[pid.codes]: http://pid.codes/1209/7D00/

### Testing procedure
1. Have a careful look at [the schematic](https://learn.adafruit.com/assets/68545) if all pins for the peripherals were chosen correctly
2. `BOARD=feather-nrf52840 make -C examples/hello-world flash term` should work with a Segger JLink Programmer connected for flashing. USB CDC ACM is used for stdio.
3. `PORT=/dev/ttyUSB0 USEMODULE=stdio_uart BOARD=feather-nrf52840 make -C examples/hello-world flash term` should work with a UART-Adapter connected.
4. `BOARD=feather-nrf52840 make -C examples/saul flash term` allows you to toggle the on-board LEDs and read out the status of the user switch.

### CODE REVIEW

**Comment by miri64** [comment]
Fixed some typos in the PR description

File: boards/feather-nrf52840/include/board.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     boards_feather-nrf52840
+ * @{
+ *
+ * @file
+ * @brief       Common board specific configuration for the Particle Mesh
```
Comment (haukepetersen) [inline boards/feather-nrf52840/include/board.h]: `Particle Mesh`? :-)

File: boards/feather-nrf52840/include/gpio_params.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    },
+    {
+        .name  = "UserSw",
+        .pin   = BTN0_PIN,
+        .mode  = BTN0_MODE,
+        .flags = SAUL_GPIO_INVERTED,
+    },
+};
+
+
```
Comment (haukepetersen) [inline boards/feather-nrf52840/include/gpio_params.h]: newline

**Comment by miri64** [comment]
Fixed.

File: boards/feather-nrf52840/board.c [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+#include "cpu.h"
+#include "board.h"
+
+#include "periph/gpio.h"
+
+void board_init(void)
+{
+    /* initialize the boards LEDs */
+    gpio_init(LED0_PIN, GPIO_OUT);
+    gpio_set(LED0_PIN);
```
Comment (haukepetersen) [inline boards/feather-nrf52840/board.c]: The LEDs are active high, and IMHO should be turned off by default (less blinding and low-power per default...). So better make this and below a `gpio_clear(x)`

**Comment by miri64** [comment]
Squashed

**Comment by miri64** [comment]
I added a workaround to allow for building with tests that

1. disable `auto_init`
2. use `ethos`

Let me know what you think @haukepetersen

**Comment by miri64** [comment]
As [discussed on devel](https://lists.riot-os.org/pipermail/devel/2020-January/006433.html) I decided to go for `stdio_rtt` for now after all due to the missing tooling around `usbus_cdc_ecm` vs. `ethos` vs. `netdev_eth` for now so we can go ahead with this PR for now.

**Comment by aabadie** [comment]
>  decided to go for stdio_rtt

Then maybe have a look at #12724: there will be no need to add the board to BOARD_BLACKLIST in different examples/tests.

**Comment by miri64** [comment]
> Then maybe have a look at #12724: there will be no need to add the board to BOARD_BLACKLIST in different examples/tests.

I know. But it is not in yet. If #12724 gets merged I remove the blacklisting of course ;-).

**Comment by aabadie** [comment]
> But it is not in yet

I know. But it's very close to be merged according to @fjmolinas's review. I would rather wait for it to be merged first if you don't mind.

**Comment by miri64** [comment]
> > But it is not in yet
> 
> I know. But it's very close to be merged according to @fjmolinas's review. I would rather wait for it to be merged first if you don't mind.

I don't mind. I think in either case the merge conflict is easily resolved :-).

File: boards/feather-nrf52840/doc.txt [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+[top-down view]: https://cdn-learn.adafruit.com/assets/assets/000/068/578/medium800/circuitpython_Screenshot_2019-01-02_at_12.04.27.png
+
+### Flash the board
+
+See the **Flashing** section in @ref boards_common_nrf52. The easiest way is to
+use an external Segger J-Link Progammer connected to the [SWD Connector].
+
+[SWD Connector]: https://learn.adafruit.com/introducing-the-adafruit-nrf52840-feather/pinouts#swd-connector-3-12
+
+### Terminal
```
Comment (miri64) [inline boards/feather-nrf52840/doc.txt]: Can someone more knowledgeable about this topic please read through this section. I'm not 100% sure I used the correct terms here :sweat_smile:

Comment (haukepetersen) [reply boards/feather-nrf52840/doc.txt]: LGTM

**Comment by miri64** [comment]
Rebased, adapted (and because of that squashed) to current master, as #12724 got merged.

**Comment by miri64** [comment]
@aabadie @haukepetersen I added another follow-up on #12724 (specifically f5cbe00118193cecedc9b3c941faab265b35e38b) to this PR, so `tests/usbus_cdc_acm_stdio` can build for the `feather-nrf52840` ([its the only thing Murdock fails on at the moment](https://ci.riot-os.org/RIOT-OS/RIOT/13085/7b3e6d35687372e6182eb64a56b40070cff5ed0f/output.html)). Let me know what you think about that.

File: Makefile.dep [hunk 5]
```diff
@@ -466,11 +466,13 @@ ifneq (,$(filter stdio_uart_rx,$(USEMODULE)))
   USEMODULE += stdio_uart
 endif
 
-ifneq (,$(filter stdio_uart,$(USEMODULE)))
-  FEATURES_REQUIRED += periph_uart
+ifneq (stdio_rtt,$(filter stdio_%,$(USEMODULE)))
+  ifneq (,$(filter stdio_%,$(USEMODULE)))
+    FEATURES_REQUIRED += periph_uart
```
Comment (aabadie) [inline Makefile.dep]: This required periph_uart in all cases except stdio_rtt so this includes `stdio_null`, `stdio_cdc_acm` but doesn't make sense.
I would decouple this:
```mk
ifneq (,$(filter stdio_uart,$(USEMODULE)))
  FEATURES_REQUIRED += periph_uart
endif

ifneq (stdio_rtt,$(filter stdio_%,$(USEMODULE)))
  # stdio_rtt cannot be used when stdio_uart/stdio_cdc_acm are loaded
  DISABLE_MODULE += stdio_rtt
endif
```

**Comment by miri64** [comment]
Last commit (5d9c9c1d97462cd0acada06abbb849afa29bdcc3) serves as a reminder to myself to remove the `stdio_cdc_acm` bullet point from the commit description of what is currently 7b3e6d35687372e6182eb64a56b40070cff5ed0f.

File: Makefile.dep (hunk 5 above)
Comment (miri64) [reply Makefile.dep]: Oops, this was not intentional.

Comment (aabadie) [reply Makefile.dep]: Re-reading my proposal, this is not better but even worse...

Comment (miri64) [reply Makefile.dep]: I fixed it as I thought it was right.

Comment (miri64) [reply Makefile.dep]: Latest commit should be the real thing now :sweat_smile:

**Comment by miri64** [comment]
The compile test results on Murdock seem weird to me. Only a few applications fail for a few `stdio_rtt`-dependent boards :-/

**Comment by miri64** [comment]
@kaspar030 @aabadie @haukepetersen can you please have a look?

**Comment by aabadie** [comment]
Maybe just add:
```
ifneq(,$(filter stdio_cdc_acm,$(USEMODULE))
    DISABLE_MODULE += stdio_rtt
endif
```
instead of the last fixup commits ?

**Comment by miri64** [comment]
That still would exclude `stdio_null` (which causes the problem for `tests/minimal` I believe) and exclude any future possible `stdio_` solution.

**Comment by aabadie** [comment]
But for now, it's fixing the broken tests for pinetime, thingy and ruuvitag. We can discuss later of better alternatives but it's out of the scope of this PR I think.

**Comment by miri64** [comment]
Ok, let's see if that works.

**Comment by miri64** [comment]
I'm away for the evening. If you want to have this in for the release, someone needs to squash this if it succeeds now.

**Comment by aabadie** [comment]
>  If you want to have this in for the release, someone needs to squash this if it succeeds now.

Makefile.dep was broken (syntax error...) with the last change. I forced push a fix, without squashing. If the CI is green and there's time for a last CI round after squashing, then this can go in the release I think.

**Comment by aabadie** [comment]
I had to push c7f502a to fix the remaining build failure on Murdock, too bad.

**Comment by fjmolinas** [comment]
Seems like this one is ready to squash? @aabadie

**Comment by aabadie** [comment]
I'm AFK. Feel free to squash @fjmolinas if you think it can go :)

**Comment by fjmolinas** [comment]
squashed.

**Comment by fjmolinas** [comment]
I trust previous reviews. GO!

**Comment by fjmolinas** [comment]
Last PR thats going in this release!

**Comment by miri64** [comment]
Thanks for reviewing and the last mile updates @fjmolinas and @aabadie!

**Pull request merged**